### PR TITLE
Do not start apps on "recompile" in an IEx session with --no-start

### LINKED ISSUE #3899: "recompile" IEx helper triggers apps start for a session with --no-start

### PR DESCRIPTION
**Pull request by lexmag**
Closes #3899.

### CODE REVIEW

File: lib/iex/lib/iex/helpers.ex [hunk 1]
```diff
@@ -80,8 +80,9 @@ defmodule IEx.Helpers do
     if mix_started? do
       config = Mix.Project.config
       reenable_tasks(config)
-      apps = stop_apps(config)
-      Mix.Task.run("app.start")
+      {not_started?, apps} = stop_apps(config)
+      args = if not_started?, do: ["--no-start"], else: []
+      Mix.Task.run("app.start", args)
       {:restarted, apps}
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: @josevalim wdyt if we rename it `restarted` –> `recompiled`?

File: lib/iex/lib/iex/helpers.ex [hunk 2]
```diff
@@ -111,8 +112,10 @@ defmodule IEx.Helpers do
         true ->
           []
       end
-    apps |> Enum.reverse |> Enum.each(&Application.stop/1)
-    apps
+    not_started? =
+      Enum.reverse(apps)
```
Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: Why you don't pipe `apps` too? :)

File: lib/iex/lib/iex/helpers.ex (hunk 1 above)
Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: Great question! Let's return accordingly to `not_started?`. The reason why I wanted to return `:restarted` was to make it clear they were restarted, so I don't want to lose this distinction. :)

Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Or maybe even better `{:ok, apps}`?

File: lib/iex/lib/iex/helpers.ex (hunk 2 above)
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: First pipe with bare variable always feels needless to me. :) You know I'm against pipelinemania like this: https://github.com/elixir-lang/ecto/pull/1040/files#diff-062a1f2a531d4698f27d0e1f61db2fbbR529.

Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: Hahaha. I hate when you have just a single pipe. :P Or you make it two, or you make it none. :P

File: lib/iex/lib/iex/helpers.ex [hunk 3]
```diff
@@ -111,8 +116,11 @@ defmodule IEx.Helpers do
         true ->
           []
       end
-    apps |> Enum.reverse |> Enum.each(&Application.stop/1)
-    apps
+    stopped? =
+      Enum.reverse(apps)
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: @josevalim and now here is two pipes. :P

Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: Evil. :P

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/iex/lib/iex/helpers.ex (hunk 3 above)
Comment (eksperimental) [inline lib/iex/lib/iex/helpers.ex]: I was thinking of writing a recent detect evry trace of evilness

Comment (eksperimental) [inline lib/iex/lib/iex/helpers.ex]: I meant REGEX not "recent"

**Pull request merged**